### PR TITLE
Add "shibaken" generator for IMDS public keys in user-data

### PR DESCRIPTION
**Pull request by jpculp**
**Issue number:**
N/A

**Description of changes:**
This lets a new user-data block be generated with public keys available
in IMDS in the event that user-data has not been populated by the user.

**Testing done:**

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

### CODE REVIEW

File: sources/api/shibaken/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+authors = ["Patrick J.P. Culp <jpculp@amazon.com>"]
+license = "Apache-2.0 OR MIT"
+edition = "2018"
+publish = false
+build = "build.rs"
+# Don't rebuild crate just because of changes to README.
+exclude = ["README.md"]
+
+[dependencies]
+apiclient = { path = "../apiclient" }
```
Comment (zmrow) [inline sources/api/shibaken/Cargo.toml]: Take a quick spin through and make sure all these dependencies are actually needed

File: sources/api/shibaken/README.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
+# shibaken
+
+Current version: 0.1.0
+
+## Introduction
+
+shibaken is called by sundog as a setting generator.
+Its sole parameter is the user-data provided by the api server.
```
Comment (zmrow) [inline sources/api/shibaken/README.md]: The parameter is the host container user data setting.

File: sources/api/shibaken/README.md [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+In the event that user-data is empty and has not been provided by the user, shibaken will generate
```
Comment (zmrow) [inline sources/api/shibaken/README.md]: Can specify `...that host container user-data is empty...`.  You probably don't need to mention that it has not been provided since the setting being empty implies that. :)

File: sources/api/shibaken/README.md [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+its own user-data by fetching authorized ssh keys from the AWS instance metadata service (IMDS).
```
Comment (zmrow) [inline sources/api/shibaken/README.md]: I think I might reword this to say:

`shibaken will fetch and populate the host container's user-data with authorized keys from ...`

File: sources/api/shibaken/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+    }
+    ensure!(
+        !public_key_endpoints.is_empty(),
+        error::NoValidImdsKeyEndpoints
+    );
+    Ok(public_key_endpoints)
+}
+
+/// Print usage message.
+fn usage() -> ! {
```
Comment (zmrow) [inline sources/api/shibaken/src/main.rs]: Let's move the usage and parse_args, down below run()

File: sources/api/shibaken/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+    let program_name = env::args().next().unwrap_or_else(|| "program".to_string());
+    eprintln!("Usage: {} user-data", program_name);
```
Comment (zmrow) [inline sources/api/shibaken/src/main.rs]: I think this is incorrect - we want the user to supply the setting  to fetch, so the invocation would look like:

`shibaken settings.host-containers.<container name>.user-data`

Take a peek at schnauzer's invocation in `sources/models/shared-defaults/defaults.toml`

File: sources/api/shibaken/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+    process::exit(2);
+}
+
+/// Parses args for the setting key name.
+fn parse_args(mut args: env::Args) -> String {
+    let arg = args.nth(1).unwrap_or_else(|| usage());
+    if arg == "--help" || arg == "-h" {
```
Comment (zmrow) [inline sources/api/shibaken/src/main.rs]: Same here - we want to receive the setting name to fetch.

File: sources/api/shibaken/src/main.rs [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+    TermLogger::init(
+        LevelFilter::Debug,
+        LogConfig::default(),
+        TerminalMode::Mixed,
+    )
+    .context(error::Logger)?;
+
+    info!("shibaken started");
+
+    info!("Fetching user-data metadata");
```
Comment (zmrow) [inline sources/api/shibaken/src/main.rs]: Fetching host-container user-data

File: sources/api/shibaken/src/main.rs [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+    let metadata_user_data = get_metadata(&setting_name, "user-data").await?;
```
Comment (zmrow) [inline sources/api/shibaken/src/main.rs]: Let's re-name this variable to just `user_data`.  `metadata_user_data` is sort of confusing.

File: sources/api/shibaken/src/main.rs [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+
+#[derive(Serialize)]
+struct Ssh {
+    authorized_keys: Vec<String>,
+}
+
+/// Returns the value of a metadata key for a given data key, erroring if the value is not a
+/// string or is empty.
+async fn get_metadata(key: &str, meta: &str) -> Result<String> {
+    let uri = &format!("{}{}?keys={}", API_METADATA_URI_BASE, meta, key);
```
Comment (zmrow) [inline sources/api/shibaken/src/main.rs]: I believe we should be querying a setting here - not metadata.  `user-data` is a setting.

File: sources/api/shibaken/src/main.rs [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+        .with_context(|| error::MetadataWrongType {
+            meta,
+            expected: "string",
+            value: response_val.to_string(),
+        })?;
+    let metadata = if !response_str.is_empty() {
+        response_str.to_string();
+    } else {
+        String::new();
+    };
```
Comment (zmrow) [inline sources/api/shibaken/src/main.rs]: I don't think we'll need to do any of this - we should be able to take the string we get back from the API, and if it's not empty, attempt to deserialize it into the structs you defined.

File: sources/api/shibaken/src/main.rs [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+    .context(error::Logger)?;
+
+    info!("shibaken started");
+
+    info!("Fetching user-data metadata");
+    let metadata_user_data = get_metadata(&setting_name, "user-data").await?;
+
+    let output = if metadata_user_data.is_empty() {
+
+        info!("No public keys available in user-data. Checking IMDS...");
```
Comment (zmrow) [inline sources/api/shibaken/src/main.rs]: This should probably read something like:

`No user-data exists for this container.  Populating user-data with keys from IMDS..."`

File: sources/api/shibaken/src/main.rs [hunk 13]
```diff
@@ -0,0 +1,318 @@
...
+        .context(error::ImdsText { uri })
+}
+
+/// Returns a list of public keys.
+fn get_public_keys_from_imds() -> Result<Vec<String>> {
+    info!("Fetching IMDS session token");
+    let client = Client::new();
+    // Use IMDSv2 for accessing instance metadata.
+    let uri = IMDS_SESSION_TOKEN_ENDPOINT;
+    let imds_session_token = client
```
Comment (zmrow) [inline sources/api/shibaken/src/main.rs]: Take a peek at `early-boot-config`'s handling of session tokens.  There, it is split into a function and the errors are a bit more descriptive.  

`sources/api/early-boot-config/src/provider/aws.rs`  < `fetch_imds_session_token()`

File: sources/api/shibaken/src/main.rs [hunk 14]
```diff
@@ -0,0 +1,318 @@
...
+        .header("X-aws-ec2-metadata-token-ttl-seconds", "60")
+        .send()
+        .context(error::ImdsRequest { method: "PUT", uri })?
+        .error_for_status()
+        .context(error::ImdsResponse { uri })?
+        .text()
+        .context(error::ImdsText { uri })?;
+
+    info!("Retrieving list of available public keys from IMDS");
+    let public_key_list =
```
Comment (zmrow) [inline sources/api/shibaken/src/main.rs:84]: Let's put a short comment here explaining what the "list" is.

File: sources/api/shibaken/src/main.rs [hunk 15]
```diff
@@ -0,0 +1,318 @@
...
+        get_text_from_imds(&client, IMDS_PUBLIC_KEY_BASE_URL, &imds_session_token)?;
+    debug!("available public keys '{}'", &public_key_list);
+
+    info!("Generating endpoint urls to retrieve text of available public keys");
```
Comment (zmrow) [inline sources/api/shibaken/src/main.rs]: This can probably be `debug!()`.  It's an implementation detail related to fetching public keys.

File: sources/api/shibaken/src/main.rs [hunk 16]
```diff
@@ -0,0 +1,318 @@
...
+    }
+    ensure!(!public_keys.is_empty(), error::NoValidImdsKeys);
+
+    Ok(public_keys)
+}
+
+/// Returns a list of endpoint urls for the public keys in IMDS. Since IMDS returns the list of
+/// available public keys as '0=my-public-key', we need to strip the index from the list and
+/// embed it into the endpoint url.
+fn get_public_key_endpoints_from_imds(public_key_list: &str) -> Result<Vec<Url>> {
```
Comment (zmrow) [inline sources/api/shibaken/src/main.rs]: At this point, we're not getting anything from IMDS.  We can probably name this function like `build_public_key_endpoints()`

File: sources/api/shibaken/src/main.rs [hunk 17]
```diff
@@ -0,0 +1,318 @@
...
+    let mut public_key_endpoints = Vec::new();
+    for available_key in public_key_list.lines() {
```
Comment (zmrow) [inline sources/api/shibaken/src/main.rs:119]: Is the output from IMDS newline delimited?

File: sources/api/shibaken/src/main.rs [hunk 18]
```diff
@@ -0,0 +1,318 @@
...
+        let f: Vec<&str> = available_key.split('=').collect();
+        // If f[0] isn't a number, then it isn't a valid index.
+        if f[0].parse::<i32>().is_ok() {
```
Comment (zmrow) [inline sources/api/shibaken/src/main.rs]: I'm still wondering if this is even possible.  It's not a bad check I suppose.

File: sources/api/shibaken/src/main.rs [hunk 19]
```diff
@@ -0,0 +1,318 @@
...
+            let public_key_endpoint = &format!("{}{}/openssh-key", IMDS_PUBLIC_KEY_BASE_URL, f[0]);
```
Comment (zmrow) [inline sources/api/shibaken/src/main.rs]: `public_key_endpoint_str`

File: sources/api/shibaken/src/main.rs [hunk 20]
```diff
@@ -0,0 +1,318 @@
...
+        let user_data_json = serde_json::to_string(&user_data).context(error::SerializeJson)?;
+        debug!("{}", &user_data_json);
+
+        info!("Encoding output");
+        // Encode the output as base64.
+        base64::encode(&user_data_json);
+
+    } else {
+        metadata_user_data;
+    };
```
Comment (zmrow) [inline sources/api/shibaken/src/main.rs]: I don't think we want this `else` condition here.  If the user-data exists, we don't need to do anything.

File: sources/api/shibaken/Cargo.toml (hunk 1 above)
Comment (jpculp) [reply sources/api/shibaken/Cargo.toml]: It looks like I need all of them except maybe http?

File: sources/api/shibaken/src/main.rs (hunk 17 above)
Comment (jpculp) [reply sources/api/shibaken/src/main.rs:119]: I believe so, but I'm not able to test it as I can't find a way to store a second key in IMDS.

File: sources/api/shibaken/src/main.rs (hunk 18 above)
Comment (jpculp) [reply sources/api/shibaken/src/main.rs]: The only way I could think of is if maybe someone put an = in their key name? Either way, doesn't hurt to check.

File: sources/api/shibaken/README.md (hunk 2 above)
Comment (jpculp) [reply sources/api/shibaken/README.md]: No longer a parameter
